dev-lead: rebase requested via @mention (on-mention intent) fails to push — non-fast-forward, no force-with-lease

## Summary

When a human asks dev-lead to rebase a PR via a free-form `@dev-lead` mention (e.g. *"@dev-lead - resolve the rebase main conflict"*), the request is classified as the **`on-mention`** intent, **not** the dedicated `rebase` intent. The engine performs the rebase (rewriting branch history), but the script's final push for `on-mention` is a **plain `git push` with no `--force-with-lease`**. Against a rebased (history-rewritten) branch that is a guaranteed **non-fast-forward rejection**, so the run fails red even though the rebase + conflict resolution succeeded.

## Evidence

Run [#27471503333](https://github.com/petry-projects/.github-private/actions/runs/27471503333/job/81203138757) — trigger: PR #624 comment `@dev-lead - resolve the rebase main conflict`.

```
INTENT_TYPE: on-mention
INTENT_REASON: human-comment-trigger
...
Everything looks correct. The rebase is clean ...
Notes: The branch is now rebased on 7390577 (main HEAD). The local branch
       diverges from origin/dev-lead/issue-573-... — the CI
       workflow will need to force-push to update the remote.
...
 ! [rejected]  dev-lead/issue-573-20260612-0205 -> dev-lead/issue-573-20260612-0205 (non-fast-forward)
error: failed to push some refs ...
##[error]git push failed — check remote access and branch permissions
##[error]Process completed with exit code 1
```

The engine explicitly noted a force-push was needed and **deferred it to the workflow** — but the `on-mention` push path does not force-push, so the rebase result is never published and the job exits 1.

## Root cause

- The dedicated **`rebase`** intent works because `prompts/dev-lead/rebase.md` (step 6) tells the engine to `git push --force-with-lease origin ${HEAD_REF}` itself, and `scripts/lib/auto-merge.sh:106-108` documents that the rebase engine force-pushes directly, bypassing `push_with_merge_guard`.
- The **`on-mention`** (and by extension `review-changes`) path calls `commit_and_push` → `push_with_merge_guard` (`scripts/lib/auto-merge.sh:122`), which runs a plain `git push` with no args. There is no force-push and no detection that the engine rewrote history.
- Result: a rebase requested via a free-form mention can never push its result.

## Key files

- `scripts/dev-lead-fix-reviews.sh` — `commit_and_push()` (~L601) and the `on-mention` case (~L977)
- `scripts/lib/auto-merge.sh` — `push_with_merge_guard()` (L122; plain `git push "$@"`)
- `prompts/dev-lead/rebase.md` — the working force-push reference (L32)

## Suggested fix (pick one, keep minimal + add tests)

1. **Detect divergence in `push_with_merge_guard`** — when local is diverged from `@{u}` (ahead AND behind), retry once with `--force-with-lease`; keep the plain push for the normal fast-forward case. This fixes every intent uniformly.
2. **Or** have `commit_and_push` pass `--force-with-lease` for intents that may rebase (`on-mention`, `review-changes`).
3. **Or** route mentions whose instruction is a rebase to the dedicated `rebase` intent in `dev-lead-intent.sh`.

Option 1 is preferred — narrowest behavioral change, covers all paths, and `--force-with-lease` stays safe against concurrent remote updates.

---
_Filed from an investigation of run #27471503333. Distinct from #573 (pr-review ci-pending re-review gap)._

--- a/scripts/lib/auto-merge.sh
+++ b/scripts/lib/auto-merge.sh
@@ -129,6 +129,27 @@ push_with_merge_guard() {
     _AM_HEAD_SHA=$(git rev-parse HEAD 2>/dev/null || true)
     return 0
   fi
+
+  # A non-fast-forward rejection while the local branch has diverged from its
+  # upstream means the engine rewrote history — e.g. it resolved a rebase under
+  # an intent (on-mention, review-changes) whose plain push can never publish
+  # the rewritten branch. Retry once with --force-with-lease, which still aborts
+  # if the remote ref advanced under us (the lease = our remote-tracking ref, so
+  # a concurrent push by someone else is never clobbered). This mirrors the
+  # dedicated rebase intent's own force-push (prompts/dev-lead/rebase.md).
+  local upstream
+  upstream=$(git rev-parse --abbrev-ref --symbolic-full-name '@{u}' 2>/dev/null || true)
+  if grep -qiE 'non-fast-forward|\[rejected\]|fetch first' "$errf" \
+     && [ -n "$upstream" ] \
+     && ! git merge-base --is-ancestor "$upstream" HEAD 2>/dev/null; then
+    echo "::notice::push rejected (non-fast-forward) and HEAD has diverged from ${upstream} — history was rewritten (likely a rebase); retrying with --force-with-lease" >&2
+    if git push --force-with-lease "$@" 2>>"$errf"; then
+      rm -f "$errf"
+      _AM_HEAD_SHA=$(git rev-parse HEAD 2>/dev/null || true)
+      return 0
+    fi
+  fi
+
   cat "$errf" >&2
   rm -f "$errf"
 

--- a/tests/dev-lead/unit/test_auto_merge.bats
+++ b/tests/dev-lead/unit/test_auto_merge.bats
@@ -185,6 +185,61 @@ EOF
   [[ "$output" == *"git push failed"* ]]
 }
 
+# ── push_with_merge_guard — force-with-lease retry on rewritten history ────────
+
+# git stub for the rebase/force-push path. $1 = "diverged" | "not-diverged"
+# controls whether merge-base --is-ancestor reports the upstream as reachable
+# from HEAD (i.e. whether history was rewritten). The plain push always reports a
+# non-fast-forward rejection; the force-with-lease push succeeds.
+_install_git_stub_force() {
+  cat > "$STUB_BIN_DIR/git" <<EOF
+#!/usr/bin/env bash
+case "\$1" in
+  push)
+    shift
+    for a in "\$@"; do
+      [ "\$a" = "--force-with-lease" ] && exit 0
+    done
+    echo "  ! [rejected]        feature -> feature (non-fast-forward)" >&2
+    exit 1
+    ;;
+  rev-parse)
+    case "\$*" in
+      *"@{u}"*) echo "origin/feature" ;;
+      *)        echo "deadbeefcafe" ;;
+    esac
+    exit 0
+    ;;
+  merge-base)
+    # --is-ancestor <upstream> HEAD: 0 = upstream reachable (not diverged),
+    # 1 = diverged (history rewritten by a rebase).
+    [ "${1:-}" = "diverged" ] && exit 1 || exit 0
+    ;;
+  *) exit 0 ;;
+esac
+EOF
+  chmod +x "$STUB_BIN_DIR/git"
+}
+
+@test "push_with_merge_guard: force-with-lease retry succeeds when history was rewritten (rebase)" {
+  source "$LIB"
+  _install_git_stub_force diverged
+  PR_STATE="open"
+  run push_with_merge_guard
+  [ "$status" -eq 0 ]
+  [[ "$output" == *"--force-with-lease"* ]]
+}
+
+@test "push_with_merge_guard: does NOT force-push when branch is not diverged (remote advanced legitimately)" {
+  source "$LIB"
+  _install_git_stub_force not-diverged
+  PR_STATE="open"
+  run push_with_merge_guard
+  [ "$status" -eq 1 ]
+  [[ "$output" == *"git push failed"* ]]
+  [[ "$output" != *"retrying with --force-with-lease"* ]]
+}
+
 # ── hold_auto_merge — commit text capture ─────────────────────────────────────
 
 @test "hold_auto_merge: captures commit title and message when auto-merge is on" {

